agent.py

```diff
@@ -440,18 +441,21 @@ def _submit_jmx_service_discovery(self, jmx_sd_configs):
             log.debug("Unable to automatically start jmxfetch on Windows via supervisor.")
 
         buffer = ""
-        for name, yaml in jmx_sd_configs.iteritems():
-            try:
+        try:
+            for name, yaml in jmx_sd_configs.iteritems():
                 buffer += SD_CONFIG_SEP
                 buffer += "# {}\n".format(name)
                 buffer += yaml
-            except Exception as e:
-                log.exception("unable to submit YAML via RPC: %s", e)
-            else:
-                log.info("JMX SD Config via named pip %s successfully.", name)
 
-        if buffer:
-            os.write(self.sd_pipe, buffer)
+            if buffer:
+                # will block if len(buffer) > OS pipe buffer.
+                # JMX will unblock when it reads on the other end.
+                os.write(self.sd_pipe, buffer)
+                os.write(self.sd_pipe, SD_CONFIG_TERM)
+        except Exception as e:
+            log.exception("unable to submit YAML via pipe: %s", e)
+        else:
+            log.info("JMX SD Config via named pipe %s successfully.", name)
```

Comment:
logging `name` here doesn't really make sense :)
Maybe you could log all the names in this entry?